src/wazuh_db/wdb_global.c

```diff
@@ -927,6 +953,74 @@ int wdb_global_delete_agent_belong(wdb_t *wdb, int id) {
     return wdb_exec_stmt_silent(stmt);
 }
 
+char *wdb_global_validate_sync_status(wdb_t *wdb, int id, const char *requested_sync_status) {
+    char *old_sync_status = wdb_global_get_sync_status(wdb, id);
+
+    if (!old_sync_status) {
+        merror("Failed to get old sync_status for agent '%d'", id);
+        // If we can't validate, allow the requested one by duplicating it
+        char *fallback = NULL;
+        os_strdup(requested_sync_status, fallback);
+        return fallback;
+    }
+
+    bool allowed = false;
+
+    if (strcmp(old_sync_status, "synced") == 0 || strcmp(old_sync_status, "syncreq_keepalive") == 0) {
+        allowed = true;
+    } else if (strcmp(old_sync_status, "syncreq_status") == 0) {
+        allowed = strcmp(requested_sync_status, "syncreq_keepalive") != 0;
+    } else if (strcmp(old_sync_status, "syncreq") == 0) {
+        allowed = strcmp(requested_sync_status, "syncreq_keepalive") != 0 &&
+                  strcmp(requested_sync_status, "syncreq_status") != 0;
+    }
```

Comment:
Perhaps just to have evidence in case of failures, we could include the case in which `old_sync_status `does not have an expected value. Returning an error message.